File: algorithm/17.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <ctype.h>
#include <math.h>

void printGrid(char **res, int size)
{
	for (int i = 0; i < size; i++)
	{
		printf("%s\n", res[i]);
	}
}

char **ans;
int ansSize;
int rear;

char dict[8][4] = {{"abc"}, {"def"}, {"ghi"}, {"jkl"}, {"mno"}, {"pqrs"}, {"tuv"}, {"wxyz"}};

void dfs(char *digits, int k, char *ele)
{
	if (k == strlen(digits))
	{
		ansSize++;
		ans = (char **)realloc(ans, ansSize * sizeof(char *));
		ans[ansSize - 1] = malloc(strlen(digits) * sizeof(char));
		memcpy(ans[ansSize - 1], ele, strlen(digits) * sizeof(char));
		return;
	}

	char *target = dict[digits[k] - '2'];
	for (int i = 0; i < strlen(target); i++)
	{
		ele[rear++] = target[i];
		dfs(digits, k + 1, ele);
		rear--;
	}
}

char **letterCombinations(char *digits, int *returnSize)
{
	ans = malloc(0);
	rear = 0;
	ansSize = 0;
	char *ele = (char *)malloc(strlen(digits) * sizeof(char));
	dfs(digits, 0, ele);
	(*returnSize) = ansSize;
	return ans;
}

int main(int argc, char *argv[])
{
	int *returnSize = (int *)calloc(1, sizeof(int));
	char digits[3] = {"23\0"};
	char **res = letterCombinations(digits, returnSize);
	printGrid(res, *returnSize);
	return 0;
}

File: go/goalgor/algor/17.go
package main

import (
	"fmt"
	"strings"
)

var dict = map[string]string{
	"2": "abc",
	"3": "def",
	"4": "ghi",
	"5": "jkl",
	"6": "mno",
	"7": "pqrs",
	"8": "tuv",
	"9": "wxyz",
}

func letterCombinations(digits string) []string {
	n := len(digits)
	if n == 0 {
		return []string{}
	}
	count := 0
	results := []string{}
	result := []string{}
	var iterate func(int)
	iterate = func(start int) {
		if count == n {
			tempResult := make([]string, len(result))
			copy(tempResult, result)
			results = append(results, strings.Join(tempResult, ""))
			return
		}
		for i := start; i < n; i++ {
			cur := dict[string(digits[i])]
			for j := 0; j < len(cur); j++ {
				result = append(result, string(cur[j]))
				count = count + 1
				iterate(i + 1)
				count = count - 1
				result = result[:len(result)-1]
			}
		}
	}
	iterate(0)
	return results
}

func main() {
	fmt.Println(letterCombinations("2"))
}
